vowpalwabbit/array_parameters.h

```diff
@@ -1,214 +1,368 @@
+	weight_map::iterator _iter;
+	uint32_t _stride;
+
+public:
+	typedef std::forward_iterator_tag iterator_category;
+	typedef T value_type;
+	typedef ptrdiff_t difference_type;
+	typedef  T* pointer;
+	typedef  T& reference;
+
+	typedef weights_iterator_iterator<T> w_iter;
+
+	sparse_weights_iterator(weight_map::iterator& iter, uint32_t stride)
+		: _iter(iter), _stride(stride)
+	{ }
+
+	sparse_weights_iterator& operator=(const sparse_weights_iterator& other)
+	{
+		_iter = other._iter;
+		_stride = other._stride;
+		return *this;
+
+	}
+	uint64_t index() { return _iter->first; }
+
+	T& operator*() { return *(_iter->second); } 
+
+	sparse_weights_iterator& operator++()
+	{  
+		_iter++;
+		return *this;
+	}
+
+	bool operator==(const sparse_weights_iterator& rhs) const { return _iter == rhs._iter; }
+	bool operator!=(const sparse_weights_iterator& rhs) const { return _iter != rhs._iter; }
+
+	//to iterate within a bucket
+	w_iter begin() { return w_iter(_iter->second);}
+	w_iter end() { return w_iter(_iter->second + _stride); }
+	w_iter end(size_t offset) { return w_iter(_iter->second + offset);}
+};
+
+
+class sparse_weight_parameters
+{
+private:
+	weight_map _map;
+	uint64_t _weight_mask;  // (stride*(1 << num_bits) -1)
+	uint32_t _stride_shift;
+	uint32_t _stride;
+	bool _seeded; // whether the instance is sharing model state with others
+	bool _delete;
+	void* default_data;
+public:
+	typedef sparse_weights_iterator<weight> iterator;
+	typedef sparse_weights_iterator<const weight> const_iterator;
+ private:
+	void(*fun)(iterator&, void*);
+ public:
+
+	sparse_weight_parameters(size_t length, uint32_t stride_shift = 0)
+		: _map(),
+		_weight_mask((length << stride_shift) - 1),
+		_stride_shift(stride_shift),
+		_stride(1 << _stride_shift),
+		_seeded(false), _delete(false),
+		fun(nullptr)
+	{}
+
+	sparse_weight_parameters()
+		: _map(), _weight_mask(0), _stride_shift(0), _stride(1), _seeded(false), _delete(false), fun(nullptr)
+	{}
+
+	bool not_null() { return (_weight_mask > 0 && !_map.empty()); }
+
+	sparse_weight_parameters(const sparse_weight_parameters &other) { shallow_copy(other); }
+	sparse_weight_parameters(sparse_weight_parameters &&) = delete;
+
+	weight* first() { throw 1; } //TODO: Throw better exceptions. Allreduce currently not supported in sparse.
+
+	//iterator with stride 
+	iterator begin() { weight_map::iterator i = _map.begin(); return iterator(i, _stride); }
+	iterator end() { weight_map::iterator i = _map.end(); return iterator(i, _stride); }
+
+	//const iterator
+	const_iterator cbegin() { weight_map::iterator i = _map.begin(); return const_iterator(i,  _stride); }
+	const_iterator cend() { weight_map::iterator i = _map.begin(); return const_iterator(i, _stride); }
+	
+	inline weight& operator[](size_t i)
+	{   uint64_t index = i & _weight_mask;
+		weight_map::iterator iter = _map.find(index);
+		if (iter == _map.end()) 
+		  {     _map.insert(std::make_pair(index, calloc_mergable_or_throw<weight>(_stride)));

```

Comment:
@eisber does mergable make sense here?  In particular, how would sharing models work when the model consists of a large number of pieces of memory?